// @mui/x-date-pickers-pro/esm/hooks/useMultiInputRangeField/useMultiInputRangeFieldSelectedSections.js
import * as React from 'react';
import useForkRef from '@mui/utils/useForkRef';
import useEventCallback from '@mui/utils/useEventCallback';
/**
 * @ignore - internal hook.
 */
export const useMultiInputRangeFieldSelectedSections = parameters => {
  const unstableEndFieldRef = React.useRef(null);
  const handleUnstableEndFieldRef = useForkRef(parameters.unstableEndFieldRef, unstableEndFieldRef);
  const [startSelectedSection, setStartSelectedSection] = React.useState(parameters.selectedSections ?? null);
  const [endSelectedSection, setEndSelectedSection] = React.useState(null);
  const getActiveField = () => {
    if (unstableEndFieldRef.current && unstableEndFieldRef.current.isFieldFocused()) {
      return 'end';
    }
    return 'start';
  };
  const handleStartSelectedSectionChange = useEventCallback(newSelectedSections => {
    setStartSelectedSection(newSelectedSections);
    if (getActiveField() === 'start') {
      parameters.onSelectedSectionsChange?.(newSelectedSections);
    }
  });
  const handleEndSelectedSectionChange = useEventCallback(newSelectedSections => {
    setEndSelectedSection(newSelectedSections);
    if (getActiveField() === 'end') {
      parameters.onSelectedSectionsChange?.(newSelectedSections);
    }
  });
  const activeField = getActiveField();
  return {
    start: {
      unstableFieldRef: parameters.unstableStartFieldRef,
      selectedSections: activeField === 'start' && parameters.selectedSections !== undefined ? parameters.selectedSections : startSelectedSection,
      onSelectedSectionsChange: handleStartSelectedSectionChange
    },
    end: {
      unstableFieldRef: handleUnstableEndFieldRef,
      selectedSections: activeField === 'end' && parameters.selectedSections !== undefined ? parameters.selectedSections : endSelectedSection,
      onSelectedSectionsChange: handleEndSelectedSectionChange
    }
  };
};
