// @mui/x-date-pickers-pro/internals/hooks/useStaticRangePicker/useStaticRangePicker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStaticRangePicker = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _PickersLayout = require("@mui/x-date-pickers/PickersLayout");
var _internals = require("@mui/x-date-pickers/internals");
var _useRangePosition = require("../useRangePosition");
var _usePickerRangePositionContext = require("../../../hooks/usePickerRangePositionContext");
var _createRangePickerStepNavigation = require("../../utils/createRangePickerStepNavigation");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["props", "steps"];
const PickerStaticLayout = (0, _styles.styled)(_PickersLayout.PickersLayout)(({
  theme
}) => ({
  overflow: 'hidden',
  minWidth: _internals.DIALOG_WIDTH,
  backgroundColor: (theme.vars || theme).palette.background.paper
}));

/**
 * Hook managing all the range static pickers:
 * - StaticDateRangePicker
 */
const useStaticRangePicker = _ref => {
  let {
      props,
      steps
    } = _ref,
    pickerParams = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  const {
    localeText,
    slots,
    slotProps,
    displayStaticWrapperAs,
    autoFocus
  } = props;
  const rangePositionResponse = (0, _useRangePosition.useRangePosition)(props);
  const getStepNavigation = (0, _createRangePickerStepNavigation.createRangePickerStepNavigation)({
    steps,
    rangePositionResponse
  });
  const {
    providerProps,
    renderCurrentView
  } = (0, _internals.usePicker)((0, _extends2.default)({}, pickerParams, {
    props,
    variant: displayStaticWrapperAs,
    autoFocusView: autoFocus ?? false,
    viewContainerRole: null,
    localeText,
    getStepNavigation
  }));
  const Layout = slots?.layout ?? PickerStaticLayout;
  const renderPicker = () => /*#__PURE__*/(0, _jsxRuntime.jsx)(_usePickerRangePositionContext.PickerRangePositionContext.Provider, {
    value: rangePositionResponse,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_internals.PickerProvider, (0, _extends2.default)({}, providerProps, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Layout, (0, _extends2.default)({}, slotProps?.layout, {
        slots: slots,
        slotProps: slotProps,
        sx: (0, _internals.mergeSx)(providerProps.contextValue.rootSx, slotProps?.layout?.sx),
        className: (0, _clsx.default)(providerProps.contextValue.rootClassName, slotProps?.layout?.className),
        ref: providerProps.contextValue.rootRef,
        children: renderCurrentView()
      }))
    }))
  });
  if (process.env.NODE_ENV !== "production") renderPicker.displayName = "renderPicker";
  return {
    renderPicker
  };
};
exports.useStaticRangePicker = useStaticRangePicker;
