// @mui/x-date-pickers-pro/DateTimeRangePicker/shared.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDateTimeRangePickerDefaultizedProps = useDateTimeRangePickerDefaultizedProps;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _styles = require("@mui/material/styles");
var _internals = require("@mui/x-date-pickers/internals");
var _hooks = require("@mui/x-date-pickers/hooks");
var _DateTimeRangePickerToolbar = require("./DateTimeRangePickerToolbar");
var _DateTimeRangePickerTabs = require("./DateTimeRangePickerTabs");
function useDateTimeRangePickerDefaultizedProps(props, name) {
  const adapter = (0, _hooks.usePickerAdapter)();
  const themeProps = (0, _styles.useThemeProps)({
    props,
    name
  });
  const validationProps = (0, _internals.useApplyDefaultValuesToDateTimeValidationProps)(themeProps);
  const ampm = themeProps.ampm ?? adapter.is12HourCycleInCurrentLocale();
  const {
    openTo,
    views: defaultViews
  } = (0, _internals.applyDefaultViewProps)({
    views: themeProps.views,
    openTo: themeProps.openTo,
    defaultViews: ['day', 'hours', 'minutes'],
    defaultOpenTo: 'day'
  });
  const {
    shouldRenderTimeInASingleColumn,
    thresholdToRenderTimeInASingleColumn,
    views,
    timeSteps
  } = (0, _internals.resolveTimeViewsResponse)({
    thresholdToRenderTimeInASingleColumn: themeProps.thresholdToRenderTimeInASingleColumn,
    ampm,
    timeSteps: themeProps.timeSteps,
    views: defaultViews
  });
  return (0, _extends2.default)({}, themeProps, validationProps, {
    timeSteps,
    openTo,
    shouldRenderTimeInASingleColumn,
    thresholdToRenderTimeInASingleColumn,
    views,
    ampm,
    slots: (0, _extends2.default)({
      tabs: _DateTimeRangePickerTabs.DateTimeRangePickerTabs,
      toolbar: _DateTimeRangePickerToolbar.DateTimeRangePickerToolbar
    }, themeProps.slots),
    slotProps: (0, _extends2.default)({}, themeProps.slotProps, {
      toolbar: (0, _extends2.default)({}, themeProps.slotProps?.toolbar, {
        ampm
      })
    })
  });
}
