// @mui/x-data-grid-premium/constants/dataGridPremiumDefaultSlotsComponents.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DATA_GRID_PREMIUM_DEFAULT_SLOTS_COMPONENTS = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _internals = require("@mui/x-data-grid-pro/internals");
var _GridPremiumColumnMenu = require("../components/GridPremiumColumnMenu");
var _material = _interopRequireDefault(require("../material"));
var _GridBottomContainer = require("../components/GridBottomContainer");
var _GridEmptyPivotOverlay = require("../components/GridEmptyPivotOverlay");
var _GridPremiumToolbar = require("../components/GridPremiumToolbar");
const DATA_GRID_PREMIUM_DEFAULT_SLOTS_COMPONENTS = exports.DATA_GRID_PREMIUM_DEFAULT_SLOTS_COMPONENTS = (0, _extends2.default)({}, _internals.DATA_GRID_PRO_DEFAULT_SLOTS_COMPONENTS, _material.default, {
  aiAssistantPanel: null,
  columnMenu: _GridPremiumColumnMenu.GridPremiumColumnMenu,
  bottomContainer: _GridBottomContainer.GridBottomContainer,
  emptyPivotOverlay: _GridEmptyPivotOverlay.GridEmptyPivotOverlay,
  toolbar: _GridPremiumToolbar.GridPremiumToolbar
});
