// @mui/x-data-grid-premium/components/promptField/PromptFieldRecord.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PromptFieldRecord = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useTimeout = require("@mui/utils/useTimeout");
var _useLazyRef = _interopRequireDefault(require("@mui/utils/useLazyRef"));
var _forwardRef = require("@mui/x-internals/forwardRef");
var _useComponentRenderer = require("@mui/x-internals/useComponentRenderer");
var _useGridRootProps = require("../../hooks/utils/useGridRootProps");
var _PromptFieldContext = require("./PromptFieldContext");
var _speechRecognition = require("../../utils/speechRecognition");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["render", "className", "onClick"];
/**
 * A button that records the user's voice when clicked.
 * It renders the `baseIconButton` slot.
 *
 * Demos:
 *
 * - [Prompt Field](https://mui.com/x/react-data-grid/components/prompt-field/)
 *
 * API:
 *
 * - [PromptFieldRecord API](https://mui.com/x/api/data-grid/prompt-field-record/)
 */
const PromptFieldRecord = exports.PromptFieldRecord = (0, _forwardRef.forwardRef)(function PromptFieldRecord(props, ref) {
  const {
      render,
      className,
      onClick
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const rootProps = (0, _useGridRootProps.useGridRootProps)();
  const {
    state,
    lang,
    onRecordingChange,
    onValueChange,
    onSubmit,
    onError
  } = (0, _PromptFieldContext.usePromptFieldContext)();
  const resolvedClassName = typeof className === 'function' ? className(state) : className;
  const recognition = (0, _useLazyRef.default)(() => {
    if (!_speechRecognition.BrowserSpeechRecognition) {
      return {
        start: () => {},
        abort: () => {}
      };
    }
    const timeout = new _useTimeout.Timeout();
    const instance = new _speechRecognition.BrowserSpeechRecognition();
    instance.continuous = true;
    instance.interimResults = true;
    instance.lang = lang;
    let finalResult = '';
    let interimResult = '';
    function start(options) {
      if (state.recording) {
        return;
      }
      onRecordingChange(true);
      instance.onresult = event => {
        finalResult = '';
        interimResult = '';
        if (typeof event.results === 'undefined') {
          instance.stop();
          return;
        }
        for (let i = event.resultIndex; i < event.results.length; i += 1) {
          if (event.results[i].isFinal) {
            finalResult += event.results[i][0].transcript;
          } else {
            interimResult += event.results[i][0].transcript;
          }
        }
        if (finalResult === '') {
          options.onUpdate(interimResult);
        }
        timeout.start(1000, () => instance.stop());
      };
      instance.onsoundend = () => {
        instance.stop();
      };
      instance.onend = () => {
        options.onDone(finalResult);
        onRecordingChange(false);
      };
      instance.onerror = error => {
        options.onError(error.message);
        instance.stop();
        onRecordingChange(false);
      };
      instance.start();
    }
    function abort() {
      instance.abort();
    }
    return {
      start,
      abort
    };
  }).current;
  const handleClick = event => {
    if (!state.recording) {
      recognition.start({
        onDone: onSubmit,
        onUpdate: onValueChange,
        onError: onError ?? (() => {})
      });
      return;
    }
    recognition.abort();
    onClick?.(event);
  };
  const element = (0, _useComponentRenderer.useComponentRenderer)(rootProps.slots.baseIconButton, render, (0, _extends2.default)({}, rootProps.slotProps?.baseIconButton, {
    className: resolvedClassName,
    disabled: state.disabled
  }, other, {
    ref,
    onClick: handleClick
  }), state);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(React.Fragment, {
    children: element
  });
});
if (process.env.NODE_ENV !== "production") PromptFieldRecord.displayName = "PromptFieldRecord";
process.env.NODE_ENV !== "production" ? PromptFieldRecord.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Override or extend the styles applied to the component.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  color: _propTypes.default.oneOf(['default', 'inherit', 'primary']),
  disabled: _propTypes.default.bool,
  edge: _propTypes.default.oneOf(['end', 'start', false]),
  id: _propTypes.default.string,
  label: _propTypes.default.string,
  /**
   * A function to customize rendering of the component.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func]),
  role: _propTypes.default.string,
  size: _propTypes.default.oneOf(['large', 'medium', 'small']),
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.number,
  title: _propTypes.default.string,
  touchRippleRef: _propTypes.default.any
} : void 0;
