// @mui/x-data-grid-premium/components/promptField/PromptFieldControl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PromptFieldControl = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _forwardRef = require("@mui/x-internals/forwardRef");
var _useComponentRenderer = require("@mui/x-internals/useComponentRenderer");
var _useGridRootProps = require("../../hooks/utils/useGridRootProps");
var _PromptFieldContext = require("./PromptFieldContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["render", "className", "onChange", "onKeyDown"];
/**
 * A component that takes user input.
 * It renders the `baseTextField` slot.
 *
 * Demos:
 *
 * - [Prompt Field](https://mui.com/x/react-data-grid/components/prompt-field/)
 *
 * API:
 *
 * - [PromptFieldControl API](https://mui.com/x/api/data-grid/prompt-field-control/)
 */
const PromptFieldControl = exports.PromptFieldControl = (0, _forwardRef.forwardRef)(function PromptFieldControl(props, ref) {
  const {
      render,
      className,
      onChange,
      onKeyDown
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const rootProps = (0, _useGridRootProps.useGridRootProps)();
  const {
    state,
    onValueChange,
    onSubmit
  } = (0, _PromptFieldContext.usePromptFieldContext)();
  const resolvedClassName = typeof className === 'function' ? className(state) : className;
  const handleChange = event => {
    onValueChange(event.target.value);
    onChange?.(event);
  };
  const handleKeyDown = event => {
    if (event.key === 'Enter' && state.value.trim()) {
      onSubmit(state.value);
    }
    onKeyDown?.(event);
  };
  const element = (0, _useComponentRenderer.useComponentRenderer)(rootProps.slots.baseTextField, render, (0, _extends2.default)({}, rootProps.slotProps?.baseTextField, {
    value: state.value,
    className: resolvedClassName
  }, other, {
    onChange: handleChange,
    onKeyDown: handleKeyDown,
    ref
  }), state);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(React.Fragment, {
    children: element
  });
});
if (process.env.NODE_ENV !== "production") PromptFieldControl.displayName = "PromptFieldControl";
process.env.NODE_ENV !== "production" ? PromptFieldControl.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  autoComplete: _propTypes.default.string,
  autoFocus: _propTypes.default.bool,
  /**
   * Override or extend the styles applied to the component.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  color: _propTypes.default.oneOf(['error', 'primary']),
  disabled: _propTypes.default.bool,
  error: _propTypes.default.bool,
  fullWidth: _propTypes.default.bool,
  helperText: _propTypes.default.string,
  id: _propTypes.default.string,
  inputRef: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.shape({
    current: _propTypes.default.object
  })]),
  label: _propTypes.default.node,
  multiline: _propTypes.default.bool,
  placeholder: _propTypes.default.string,
  /**
   * A function to customize rendering of the component.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func]),
  role: _propTypes.default.string,
  size: _propTypes.default.oneOf(['medium', 'small']),
  slotProps: _propTypes.default.object,
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.number,
  type: _propTypes.default.oneOfType([_propTypes.default.oneOf(['button', 'checkbox', 'color', 'date', 'datetime-local', 'email', 'file', 'hidden', 'image', 'month', 'number', 'password', 'radio', 'range', 'reset', 'search', 'submit', 'tel', 'text', 'time', 'url', 'week']), _propTypes.default.object]),
  value: _propTypes.default.string
} : void 0;
