// @mui/x-data-grid-premium/material/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _xDataGrid = require("@mui/x-data-grid");
var _xDataGridPro = require("@mui/x-data-grid-pro");
var _icons = require("./icons");
const iconsSlots = {
  collapsibleIcon: _xDataGrid.GridExpandMoreIcon,
  columnMenuUngroupIcon: _icons.GridWorkspacesIcon,
  columnMenuGroupIcon: _icons.GridGroupWorkIcon,
  columnMenuAggregationIcon: _icons.GridFunctionsIcon,
  pivotIcon: _icons.GridPivotIcon,
  pivotSearchIcon: _xDataGrid.GridSearchIcon,
  pivotSearchClearIcon: _xDataGrid.GridClearIcon,
  pivotMenuAddIcon: _xDataGrid.GridAddIcon,
  pivotMenuMoveUpIcon: _icons.GridExpandLessIcon,
  pivotMenuMoveDownIcon: _xDataGrid.GridExpandMoreIcon,
  pivotMenuMoveToTopIcon: _icons.GridMoveToTopIcon,
  pivotMenuMoveToBottomIcon: _icons.GridMoveToBottomIcon,
  pivotMenuCheckIcon: _xDataGrid.GridCheckIcon,
  pivotMenuRemoveIcon: _xDataGrid.GridDeleteIcon,
  sidebarCloseIcon: _xDataGrid.GridCloseIcon,
  aiAssistantIcon: _icons.GridAssistantIcon,
  aiAssistantPanelCloseIcon: _xDataGrid.GridCloseIcon,
  aiAssistantPanelNewConversationIcon: _xDataGrid.GridAddIcon,
  aiAssistantPanelHistoryIcon: _icons.GridHistoryIcon,
  promptIcon: _icons.GridPromptIcon,
  promptSendIcon: _icons.GridSendIcon,
  promptSpeechRecognitionIcon: _icons.GridMicIcon,
  promptSpeechRecognitionOffIcon: _icons.GridMicOffIcon,
  promptRerunIcon: _icons.GridRerunIcon,
  promptSortAscIcon: _xDataGridPro.GridArrowUpwardIcon,
  promptSortDescIcon: _xDataGridPro.GridArrowDownwardIcon,
  promptFilterIcon: _xDataGridPro.GridFilterAltIcon,
  promptPivotIcon: _icons.GridPivotIcon,
  promptAggregationIcon: _icons.GridFunctionsIcon,
  promptGroupIcon: _icons.GridGroupWorkIcon,
  promptChangesToggleIcon: _xDataGrid.GridExpandMoreIcon
};
const materialSlots = (0, _extends2.default)({}, iconsSlots);
var _default = exports.default = materialSlots;
