// @mui/x-data-grid-pro/hooks/features/treeData/gridTreeDataGroupColDef.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GRID_TREE_DATA_GROUPING_COL_DEF_FORCED_PROPERTIES = exports.GRID_TREE_DATA_GROUPING_COL_DEF = void 0;
Object.defineProperty(exports, "GRID_TREE_DATA_GROUPING_FIELD", {
  enumerable: true,
  get: function () {
    return _internals.GRID_TREE_DATA_GROUPING_FIELD;
  }
});
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _xDataGrid = require("@mui/x-data-grid");
var _internals = require("@mui/x-data-grid/internals");
/**
 * TODO: Add sorting and filtering on the value and the filteredDescendantCount
 */
const GRID_TREE_DATA_GROUPING_COL_DEF = exports.GRID_TREE_DATA_GROUPING_COL_DEF = (0, _extends2.default)({}, _xDataGrid.GRID_STRING_COL_DEF, {
  type: 'custom',
  sortable: false,
  filterable: false,
  disableColumnMenu: true,
  disableReorder: true,
  align: 'left',
  width: 200,
  valueGetter: (value, row, column, apiRef) => {
    const rowId = (0, _xDataGrid.gridRowIdSelector)(apiRef, row);
    const rowNode = (0, _xDataGrid.gridRowNodeSelector)(apiRef, rowId);
    return rowNode?.type === 'group' || rowNode?.type === 'leaf' ? rowNode.groupingKey : undefined;
  }
});
const GRID_TREE_DATA_GROUPING_COL_DEF_FORCED_PROPERTIES = exports.GRID_TREE_DATA_GROUPING_COL_DEF_FORCED_PROPERTIES = {
  field: _internals.GRID_TREE_DATA_GROUPING_FIELD,
  editable: false,
  groupable: false
};
